src/libappimage/desktop_integration/Thumbnailer.cpp

```diff
@@ -0,0 +1,156 @@
+#include "utils/IconHandle.h"
+#include "Thumbnailer.h"
+
+using namespace appimage::utils;
+namespace bf = boost::filesystem;
+
+namespace appimage {
+    namespace desktop_integration {
+        Thumbnailer::Thumbnailer() : xdgCacheHome(XdgUtils::BaseDir::Home() + "/.cache") {}
+
+        Thumbnailer::Thumbnailer(const std::string& xdgCacheHome) : xdgCacheHome(xdgCacheHome) {
+            if (Thumbnailer::xdgCacheHome.empty())
+                Thumbnailer::xdgCacheHome = XdgUtils::BaseDir::Home() + "/.cache";
+        }
+
+        void Thumbnailer::create(const core::AppImage& appImage) {
+            auto resources = extractResources(appImage);
+            std::string appIcon = getAppIconName(resources);
+
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+
+            auto normalIconData = getIconData(resources, appIcon, "128x128");
+            generateNormalSizeThumbnail(canonicalPathMd5, normalIconData);
+
+            auto largeIconData = getIconData(resources, appIcon, "256x256");
+            generateLargeSizeThumbnail(canonicalPathMd5, largeIconData);
+        }
+
+        void Thumbnailer::remove(const core::AppImage& appImage) {
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+
+            bf::remove(normalThumbnailPath);
+            bf::remove(largeThumbnailPath);
+        }
+
+        void Thumbnailer::generateNormalSizeThumbnail(const std::string& canonicalPathMd5,
+                                                      std::vector<char>& normalIconData) const {
+
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::create_directories(normalThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{normalIconData.begin(), normalIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(128);
+                iconHandle.save(normalThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(normalThumbnailPath);
+            out.write(normalIconData.data(), normalIconData.size());
+        }
+
+        void Thumbnailer::generateLargeSizeThumbnail(const std::string& canonicalPathMd5,
+                                                     std::vector<char>& largeIconData) const {
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+            bf::create_directories(largeThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{largeIconData.begin(), largeIconData.end()};
```

Comment:
This use will create an empty vector, then call `std::copy` and cause a lot of allocations therefore. Why not simply use the copy constructor?

```
std::vector<uint8_t> iconData(largeIconData);
```

This will allocate enough memory before copying the vector.

In this special case, you could also just pass by value, which will take care of the copying. But I assume you want to avoid unnecessary copying when there's exceptions from the fs operations. Needed to look twice to recognize that, a comment might help.

This comment applies to all occurrences of this snippet in the file.

Reply:
It's not possible to directly assign a `std::vector<char>` to a `std::vector<uint8_t>`. The transformation is required by the `librsvg` and `libacairo` functions. I think that there's room for improvements in the `IconHandle` class. But they aren't a priority right now.

Reply:
Still, specifying a size will pre-allocate enough memory.

Why bother using `uint8_t` anyway? `char` has the same size, and is used for bytes in almost all (especially C) code.
You could also perform a `reinterpret_cast`, it's safe to use it here.

Reply:
We can save the cast to `uint8_t` until the last time and save a few copies. Changes applied.